proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,264 @@
+        { i32 2, i32 0, i32 200 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 3, i32 0, i32 212 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 4, i32 0, i32 224 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 0, i32 0, i32 236 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 0, i32 0, i32 316 }, 
+        %"dxbc::RootSignature::ContainerRootParameter" 
+        { i32 1, i32 0, i32 348 }], 
+      [2 x %"dxbc::RootSignature::StaticSamplerDesc"]
+         [%"dxbc::RootSignature::StaticSamplerDesc" 
+           { i32 85, i32 1, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, 
+             i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 1, 
+             i32 0, i32 0 }, 
+          %"dxbc::RootSignature::StaticSamplerDesc" 
+           { i32 21, i32 3, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, 
+           i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 2, 
+           i32 0, i32 0 }], 
+      %1 { 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          { i32 0, i32 1, i32 8 }, 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          zeroinitializer, 
+        %"dxbc::RootSignature::ContainerRootDescriptor" 
+          zeroinitializer, 
+        %2 { %"dxbc::RootSignature::ContainerRootDescriptorTable" 
+        { i32 3, i32 244 }, 
+        [3 x %"dxbc::RootSignature::ContainerDescriptorRange"] 
+         [%"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 2, i32 1, i32 1, i32 0, i32 0, i32 -1 }, 
+          %"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 0, i32 8, i32 1, i32 0, i32 1, i32 -1 }, 
+          %"dxbc::RootSignature::ContainerDescriptorRange" 
+           { i32 1, i32 -1, i32 1, i32 0, i32 1, i32 -1 }] 
+       }, 
+       %3 {
+           %"dxbc::RootSignature::ContainerRootDescriptorTable" 
+           { i32 1, i32 324 }, 
+           [1 x %"dxbc::RootSignature::ContainerDescriptorRange"] 
+           [%"dxbc::RootSignature::ContainerDescriptorRange" 
+             { i32 3, i32 4, i32 0, i32 1, i32 0, i32 -1 }] }, 
+           %"dxbc::RootSignature::ContainerRootConstants" 
+             { i32 10, i32 0, i32 3 } 
+       }
+      }
+    }
+
+```
+
+* LLVM IR for root signature saved in metadata.
+
+```
+
+  ; named metadata for entry root signature
+  !hlsl.entry.rootsignatures = !{!3}
+  …
+  !3 = !{ptr @main, !4}
+  !4 = !{i32 2, i32 3, !5, !18}      ; version 1.1, flag 3, root params, statis samplers.
+  !5 = !{!6, !7, !8, !9, !14, !17}  ; root parameters.
+  !6 = !{i32 2, i32 0, i32 0, i32 1, i32 8}
+  !7 = !{i32 3, i32 0, i32 0, i32 0, i32 0}
+  !8 = !{i32 4, i32 0, i32 0, i32 0, i32 0}
+  !9 = !{i32 0, i32 0, !10}  ; first table
+  !10 = !{!11, !12, !13} ; table ranges.
+  !11 = !{i32 2, i32 1, i32 1, i32 0, i32 0}
+  !12 = !{i32 0, i32 8, i32 1, i32 0, i32 1}
+  !13 = !{i32 1, i32 -1, i32 1, i32 0, i32 1}
```

Comment:
If the last value in the tuple is OffsetInDescriptorsFromTableStart, then shouldn't the one in `!13` be `9` - so it's after the previous two ranges (1 + 8 = 9).

Reply:
Or maybe it should be "-1" to indicate "DESCRIPTOR_RANGE_OFFSET_APPEND".  I'm not sure if your design is to calculate the actual descriptor range offsets (which is implied by these having the value "1" in here) or if it is to just represent what was specified in the HLSL source (which would be the default value for OffsetInDescriptorsFromTableStart = DESCRIPTOR_RANGE_OFFSET_APPEND = -1 in this case).